app/routes/surveys/components/SurveyEditor/SurveyEditor.js

```diff
@@ -0,0 +1,189 @@
+import { Field, FieldArray, formValueSelector } from 'redux-form';
+import { connect } from 'react-redux';
+import Button from 'app/components/Button';
+import {
+  TextInput,
+  CheckBox,
+  SelectInput,
+  DatePicker,
+  legoForm
+} from 'app/components/Form';
+import { createValidator, required } from 'app/utils/validation';
+import type { SurveyEntity } from 'app/reducers/surveys';
+import { Content } from 'app/components/Content';
+import type { FieldProps } from 'redux-form';
+import Icon from 'app/components/Icon';
+import { Link } from 'react-router';
+
+type Props = FieldProps & {
+  survey: SurveyEntity,
+  autoFocus: Object,
+  surveyData: Array<Object>,
+  submitFunction: (SurveyEntity, ?number) => Promise<*>,
+  deleteSurvey: number => Promise<*>,
+  push: string => void
+};
+
+const SurveyEditor = ({
+  survey,
+  submitting,
+  autoFocus,
+  handleSubmit,
+  deleteSurvey,
+  questions
+}: Props) => {
+  const titleField = (
+    <Field
+      placeholder="Tittel"
+      label=" "
+      autoFocus={autoFocus}
+      name="title"
+      component={TextInput.Field}
+      className={styles.editTitle}
+    />
+  );
+
+  return (
+    <Content className={styles.detail}>
+      <form onSubmit={handleSubmit}>
+        {survey && survey.id ? (
+          <DetailNavigation
+            title={titleField}
+            surveyId={Number(survey.id)}
+            deleteFunction={deleteSurvey}
+          />
+        ) : (
+          <ListNavigation title={titleField} />
+        )}
+
+        <div className={styles.checkBox}>
+          <Field
+            name="isClone"
+            label="Klone av en annen undersøkelse?"
+            component={CheckBox.Field}
+            normalize={v => !!v}
+          />
+        </div>
+
+        <div style={{ display: 'flex', justifyContent: 'space-between' }}>
+          <Field
+            placeholder="Bekk Miniseminar"
+            label="Arrangement"
+            autoFocus={autoFocus}
+            name="event"
+            component={SelectInput.AutocompleteField}
+            className={styles.editEvent}
+            filter={['events.event']}
+          />
+
+          <Field
+            label="Aktiveringstidspunkt"
+            name="activeFrom"
+            component={DatePicker.Field}
+          />
+        </div>
+
+        <FieldArray
+          name="questions"
+          questions={questions}
+          component={renderQuestions}
+        />
+
+        <div className={styles.clear} />
+        <Button className={styles.submit} disabled={submitting} submit>
+          Lagre
+        </Button>
+      </form>
+    </Content>
+  );
+};
+
+const renderQuestions = ({ fields, meta: { touched, error }, questions }) => {
+  return [
+    <ul className={styles.questions} key="questions">
+      {fields.map((question, i) => (
+        <Question
+          key={i}
+          index={i}
+          question={question}
+          questionData={questions ? questions[i] : {}}
```

Comment:
ugh. stupid `FieldArray`. Could we perhaps make it more clear what `question` is here? `questionId` or whatever is it?

Reply:
Yeah, this naming schema isn't ideal. The `questionData` comes from here: https://github.com/webkom/lego-webapp/pull/1119/files#diff-abb76d391c7bf50ade8ddf7cd0d08d97R185 (and forwarded to `renderQuestions` here https://github.com/webkom/lego-webapp/pull/1119/files#diff-abb76d391c7bf50ade8ddf7cd0d08d97R94), and is selected from the `redux(-form)` state. We have to do that, since we need the `questionType` inside of the `QuestionComponent`.

The `question` prop is just the form-name that the current `question` is given.

Reply:
I guess it would be a bit easier to read (and understand) with more flow entities :smile: Redux-form v7 will also help, with way better flow support.